### DISCUSSION TITLE
Support multi language

### DISCUSSION THREAD

**Discussion by mauroerta** [mauroerta/deno-plural#7] [Ideas]
As suggested by @vixalien in [this](https://github.com/mauroerta/deno-plural/pull/5#issuecomment-1107441289) comment, it could be nice for `deno-plural` to support other languages other than English.
This discussion was open to finding the proper way to implement this behavior.

Any proposal or idea is welcome!
I'll comment soon with mine.

**Comment by vixalien** [comment]
Hey @mauroerta I was thinking of a thing like the following:

-  First move the "data" file to another directory such as "rules/english".
- Create rules for languages we understand first by looking for other projects or through Wikipedia pages or whatnot. I could help with French if possible.

RFC

**Reply by mauroerta** [reply]
First of all, I'm really sorry for answering only now.
About your idea, I think it makes sense. It'd be pretty easy for now to move just the English rules in the `rules` folder and then, whenever we understand how to add a new language, create another file under the `rules` folder.

The folder structure, if I got it right, would be something like this:

```console
.
└── src/
    ├── rules/
    │   ├── en.ts
    │   ├── fr.ts
    │   ├── it.ts
    │   └── ...anyOtherSupportedLanguage.ts
    ├── denoPlural.ts
    ├── denoPlural_tests.ts
    └── mod.ts
```

What we are missing I think is just a way to initialize the package in order to retrieve the right rules, something like:

```typescript
import { setLanguage, plural } from "https://deno.land/x/deno_plural/mod.ts";

setLanguage('en');

console.log(plural('car')) // --> cars

setLanguage('fr');

console.log(plural('auto')) // --> voitures

setLanguage('it');

console.log(plural('automobile')) // --> automobili

```

I think it can also be nice if we can specify the language `inline` as a second parameter of the functions `plural` and `singular`:

```typescript
import { setLanguage, singular, plural } from "https://deno.land/x/deno_plural/mod.ts";

setLanguage('en');
console.log(singular('cars')) // --> car `default is the current languagee`
console.log(singular('voitures', 'fr')) // --> auto
console.log(plural('automobile', 'it')) // --> automobili
```

What do you think about it?

> PS since I don't speak french, in case `voitures` is not the plural of `auto` let me know and I'll update this comment

**Comment by vixalien** [comment]
Hey @mauroerta 

I believe it may be a good idea to allow the user to load their rules (even from their own sources) instead of loading all of the rules at once as they may be many and large. I assume English will be loaded and `set` by default.

Here is a loose implementation based on the syntax I used in [written](https://github.com/written).

```ts
import { plural, setLanguage } as written from "https://deno.land/x/deno_plural/mod.ts";
import FR from "https://deno.land/x/deno_plural/rules/fr.ts";
// there is also es,fs,it and se.

// the second parameter may be necessary to allow rules to be identified
setLanguage(FR, "fr");

// then use it like so
plural("voiture"); // voitures
```

I completely agree with you on the second parameter on `plural` so as to use a different language inline.

**Reply by mauroerta** [reply]
Cool, I like it!
I'll try to provide an implementation for this as soon as possible.

**Reply by mauroerta** [reply]
The implementation I'm proposing in [this PR](https://github.com/mauroerta/deno-plural/pull/8) is slightly different:

## To add new rules:

```typescript
// addRules('LANG_KEY', RULES);
addRules('fr', frRules);
```

## Change the language:

Once the language is set, we can easily set it as the current language:

```typescript
// setLanguage('LANG_KEY');
setLanguage('fr');
```

And then use it with the functions `plural` and `singular`:

```typescript
plural('voiture'); // voitures
plural('car', 'en'); // cars <-- new language applied inline only for this translation
singular('voitures'); // voiture
```

**Reply by mauroerta** [reply]
Let me know what do you think about it, in case I'll merge it and publish a new version!

**Comment by mauroerta** [comment]
@vixalien I opened a [PR](https://github.com/mauroerta/deno-plural/pull/8) with the changes, whenever you can be really helpful if you can take a look and let me know!

**Reply by mauroerta** [reply]
@vixalien The PR has been merged, soon I'll publish a new version

**Reply by vixalien** [reply]
Good work @mauroerta
